File: MinFallingPathSum.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace LeetCodeSuper30
{
    class MinFallingPathSumLC
    {
        //Time Complexity: O(m*n)
        //Space Complexity: O(1)
        public int MinFallingPathSum(int[][] matrix)
        {
            if (matrix == null || matrix.GetLength(0) == 0)
            {
                return 0;
            }
            int m = matrix.GetLength(0);
            int n = matrix.GetLength(1);
            for (int i = 1; i < m; i++)
            {
                for (int j = 0; j < n; j++)
                {
                    if (j == 0)
                    {
                        matrix[i][j] += Math.Min(matrix[i - 1][j], matrix[i - 1][j + 1]);
                    }
                    else if (j == n - 1)
                    {
                        matrix[i][j] += Math.Min(matrix[i - 1][j], matrix[i - 1][j - 1]);
                    }
                    else
                    {
                        matrix[i][j] += Math.Min(matrix[i - 1][j], Math.Min(matrix[i - 1][j + 1], matrix[i - 1][j - 1]));
                    }
                }
            }
            int min = Int32.MaxValue;
            for (int i = 0; i < n; i++)
            {
                min = Math.Min(min, matrix[m - 1][i]);
            }
            return min;
        }
    }
}

File: assignments/src/Min_falling_path_sum.java
public class Min_falling_path_sum {

    // Time: O(n^2), Space: O(n)
    public int minFallingPathSum(int[][] matrix) {

        int dp[] = new int[matrix[0].length+1];

        for (int i = matrix.length-1; i>=0; i--) {
            int currRow[] = new int[matrix[0].length+1];
            for (int j = 0; j<matrix.length; j++) {
                if (j == 0) {
                    currRow[j] = Math.min(dp[j], dp[j+1]) + matrix[i][j];
                }
                else if (j == matrix.length-1) {
                    currRow[j] = Math.min(dp[j], dp[j-1]) + matrix[i][j];
                }
                else {
                    currRow[j] = Math.min(dp[j], Math.min(dp[j+1], dp[j-1])) + matrix[i][j];
                }
            }
            dp = currRow;
        }
        int min = Integer.MAX_VALUE;
        for (int i = 0; i<matrix.length; i++) {
            if (dp[i] < min) min = dp[i];
        }
        return min;
    }
}